src/System.Private.CoreLib/shared/System/MemoryExtensions.cs

```diff
@@ -1607,5 +1607,145 @@ public static bool Overlaps<T>(this ReadOnlySpan<T> span, ReadOnlySpan<T> other,
+        }
+
+        public static void Sort<TKey, TValue>(this Span<TKey> keys, Span<TValue> items)
+        {
+            if (keys == null)
+                ThrowHelper.ThrowArgumentNullException(ExceptionArgument.keys);
+            Sort<TKey, TValue>(keys!, items, 0, keys!.Length, null); // TODO-NULLABLE: https://github.com/dotnet/csharplang/issues/538
+        }
+
+        public static void Sort<T>(this Span<T> array, int index, int length)
+        {
+            Sort<T>(array, index, length, null);
+        }
+
+        public static void Sort<TKey, TValue>(this Span<TKey> keys, Span<TValue> items, int index, int length)
+        {
+            Sort<TKey, TValue>(keys, items, index, length, null);
+        }
+
+        public static void Sort<T>(this Span<T> array, System.Collections.Generic.IComparer<T>? comparer)
+        {
+            if (array == null)
+                ThrowHelper.ThrowArgumentNullException(ExceptionArgument.array);
+            Sort<T>(array!, 0, array!.Length, comparer); // TODO-NULLABLE: https://github.com/dotnet/csharplang/issues/538
+        }
+
+        public static void Sort<TKey, TValue>(this Span<TKey> keys, Span<TValue> items, System.Collections.Generic.IComparer<TKey>? comparer)
+        {
+            if (keys == null)
+                ThrowHelper.ThrowArgumentNullException(ExceptionArgument.keys);
+            Sort<TKey, TValue>(keys!, items, 0, keys!.Length, comparer); // TODO-NULLABLE: https://github.com/dotnet/csharplang/issues/538
+        }
+
+        public static void Sort<T>(this Span<T> array, int index, int length, System.Collections.Generic.IComparer<T>? comparer)
+        {
+            if (array == null)
+                ThrowHelper.ThrowArgumentNullException(ExceptionArgument.array);
+            if (index < 0)
+                ThrowHelper.ThrowIndexArgumentOutOfRange_NeedNonNegNumException();
+            if (length < 0)
+                ThrowHelper.ThrowLengthArgumentOutOfRange_ArgumentOutOfRange_NeedNonNegNum();
+            if (array!.Length - index < length) // TODO-NULLABLE: https://github.com/dotnet/csharplang/issues/538
+                ThrowHelper.ThrowArgumentException(ExceptionResource.Argument_InvalidOffLen);
+
+            if (length > 1)
+            {
+#if CORECLR
+                if (comparer == null || comparer == Comparer<T>.Default)
+                {
+                    // Array based
+                    //if (TrySZSort(array, null, index, index + length - 1))
+                    //{
+                    //    return;
+                    //}
+                    unsafe
+                    {
+                        ref byte byteRef = ref Unsafe.As<T, byte>(ref array.GetPinnableReference());
+                        fixed (byte* ptr = &byteRef)
+                        {
+                            if (Array.TrySZSort(new IntPtr(ptr), typeof(T), IntPtr.Zero, typeof(T), index, index + length - 1))
```

Comment:
Need a way to either call `TrySZSort` here for spans as suggested here, or spans can revert to slower managed code.